src/arithmetic.cpp

```diff
@@ -383,40 +116,15 @@ void _lsubtract(fp* z, const fp* x, const fp* y)
 #endif
 
 #ifdef __x86_64__
-void __negate(fp* z, const fp* x)
-{
-    // x86_64 calling convention (https://en.wikipedia.org/wiki/X86_calling_conventions#System_V_AMD64_ABI):
-    // z => %rdi
-    // x => %rsi
-    // callee needs to restore registers r15, r14, r13, r12, rbx before returning
-    asm("mov $0xb9feffffffffaaab,%r8;");
-    asm("mov $0x1eabfffeb153ffff,%r9;");
-    asm("mov $0x6730d2a0f6b0f624,%r10;");
-    asm("mov $0x64774b84f38512bf,%r11;");
-    asm("mov $0x4b1ba7b6434bacd7,%rcx;");
-    asm("mov $0x1a0111ea397fe69a,%rax;");
-    asm("sub     (%rsi),%r8;");
-    asm("sbb 0x08(%rsi),%r9;");
-    asm("sbb 0x10(%rsi),%r10;");
-    asm("sbb 0x18(%rsi),%r11;");
-    asm("sbb 0x20(%rsi),%rcx;");
-    asm("sbb 0x28(%rsi),%rax;");
-    asm("mov %r8,     (%rdi);");
-    asm("mov %r9, 0x08(%rdi);");
-    asm("mov %r10,0x10(%rdi);");
-    asm("mov %r11,0x18(%rdi);");
-    asm("mov %rcx,0x20(%rdi);");
-    asm("mov %rax,0x28(%rdi);");
-}
+void __negate(fp* z, const fp* x);
 void _negate(fp* z, const fp* x)
 {
-    __negate(z, x);
-    // put zero check after __neg because gcc messes up %rdi in -O3 (doesn't restore it before inlining asm code)
```

Comment:
This is one area that indicates we were running afoul of `asm()` rules. I refactored this to do the `if(x->isZero())` before calling `__negate()` so it matches that as the C impl below